jetty-security/src/main/java/org/eclipse/jetty/security/authentication/SpnegoAuthenticator2.java

```diff
@@ -0,0 +1,218 @@
+    public String getAuthMethod()
+    {
+        return _authMethod;
+    }
+
+    public Duration getAuthenticationDuration()
+    {
+        return _authenticationDuration;
+    }
+
+    /**
+     * <p>Sets the duration of the authentication.</p>
+     * <p>A negative duration means that the authentication is only valid for the current request.</p>
+     * <p>A zero duration means that the authentication is valid forever.</p>
+     * <p>A positive value means that the authentication is valid for the specified duration.</p>
+     *
+     * @param authenticationDuration the authentication duration
+     */
+    public void setAuthenticationDuration(Duration authenticationDuration)
+    {
+        _authenticationDuration = authenticationDuration;
+    }
+
+    @Override
+    public Authentication validateRequest(ServletRequest req, ServletResponse res, boolean mandatory) throws ServerAuthException
+    {
+        if (!mandatory)
+            return new DeferredAuthentication(this);
+
+        HttpServletRequest request = (HttpServletRequest)req;
+        HttpServletResponse response = (HttpServletResponse)res;
+
+        HttpSession httpSession = request.getSession(false);
+        if (httpSession != null)
+        {
+            UserIdentityHolder holder = (UserIdentityHolder)httpSession.getAttribute(UserIdentityHolder.ATTRIBUTE);
+            if (holder != null)
+            {
+                UserIdentity identity = holder._userIdentity;
+                if (identity != null)
+                {
+                    Duration authnDuration = getAuthenticationDuration();
+                    if (!authnDuration.isNegative())
+                    {
+                        boolean expired = !authnDuration.isZero() && Instant.now().isAfter(holder._validFrom.plus(authnDuration));
+                        // Allow non-GET requests even if they're expired, so that
+                        // the client does not need to send the request content again.
+                        if (!expired || !HttpMethod.GET.is(request.getMethod()))
+                            return new UserAuthentication(getAuthMethod(), identity);
+                    }
+                }
+            }
+        }
+
+        String header = request.getHeader(HttpHeader.AUTHORIZATION.asString());
+        String spnegoToken = getSpnegoToken(header);
+
+        // The client has responded to the challenge we sent previously.
+        if (header != null && spnegoToken != null)
+        {
+            SpnegoUserIdentity identity = (SpnegoUserIdentity)login(null, spnegoToken, request);
+            if (identity != null)
+            {
+                if (identity.isEstablished())
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Sending final challenge");
+                    SpnegoUserPrincipal principal = (SpnegoUserPrincipal)identity.getUserPrincipal();
+                    setSpnegoToken(response, principal.getEncodedToken());
```

Comment:
Do you need to check if this is a deferred auth here? if so, you don't real response at this point.  You can probably still authenticate... maybe... but can't set the token in the response... is that fatal?